File: P86.java
import java.util.Stack;

/**
 * Created on 19/04/2019.
 * Description:
 *
 * @author iznauy
 */
public class P86 {

    private static final class ListNode {
        int val;
        ListNode next;
        ListNode(int x) { val = x; }
    }

    public ListNode partition(ListNode head, int x) {
        Stack<ListNode> stack1 = new Stack<>();
        Stack<ListNode> stack2 = new Stack<>();
        ListNode curr = head;
        while (curr != null) {
            if (curr.val < x)
                stack1.push(curr);
            else
                stack2.push(curr);
            curr = curr.next;
        }
        while (!stack2.empty()) {
            ListNode temp = stack2.pop();
            temp.next = curr;
            curr = temp;
        }
        while (!stack1.empty()) {
            ListNode temp = stack1.pop();
            temp.next = curr;
            curr = temp;
        }
        return curr;
    }
}

File: p86.cpp
//
// Created by iznauy on 19/04/2019.
//
#include <bits/stdc++.h>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode* partition(ListNode* head, int x) {
        auto dummy_one = new ListNode(0);
        auto dummy_two = new ListNode(0);
        ListNode * curr_one = dummy_one;
        ListNode * curr_two = dummy_two;

        ListNode * curr = head;
        while (curr != NULL) {
            if (curr->val < x) {
                curr_one->next = curr;
                curr_one = curr_one->next;
            } else {
                curr_two->next = curr;
                curr_two = curr_two->next;
            }
            curr = curr->next;
        }
        curr_one->next = dummy_two->next;
        curr_two->next = NULL;
        return dummy_one->next;
    }
};